File: julia/src/benchmarks/RegexDna.jl
mutable struct RegexDna <: AbstractBenchmark
    seq::String
    ilen::Int32
    clen::Int32
    io::IOBuffer
    fasta_n::Int64

    function RegexDna()
        fasta_n = Helper.config_i64("RegexDna", "n")
        new("", Int32(0), Int32(0), IOBuffer(), fasta_n)
    end
end

name(b::RegexDna)::String = "RegexDna"

function prepare(b::RegexDna)

    fasta = Fasta()
    fasta.n = b.fasta_n
    run(fasta, 0)

    if isdefined(fasta, :io)
        seekstart(fasta.io)
        res = String(take!(fasta.io))
    elseif isdefined(fasta, :result)
        res = fasta.result
    else
        error("Fasta должен иметь поле io или result")
    end

    b.ilen = Int32(length(res))

    seq_chunks = String[]
    for line in split(res, '\n')
        if !startswith(line, '>')
            push!(seq_chunks, line)
        end
    end

    b.seq = join(seq_chunks)
    b.clen = Int32(sizeof(b.seq))
end

function run(b::RegexDna, iteration_id::Int64)
    io = b.io

    patterns = [
        r"agggtaaa|tttaccct",
        r"[cgt]gggtaaa|tttaccc[acg]",
        r"a[act]ggtaaa|tttacc[agt]t",
        r"ag[act]gtaaa|tttac[agt]ct",
        r"agg[act]taaa|ttta[agt]cct",
        r"aggg[acg]aaa|ttt[cgt]ccct",
        r"agggt[cgt]aa|tt[acg]accct",
        r"agggta[cgt]a|t[acg]taccct",
        r"agggtaa[cgt]|[acg]ttaccct",
    ]

    for regex in patterns
        count = 0
        pos = 1
        while true
            m = match(regex, b.seq, pos)
            if m === nothing
                break
            end
            count += 1
            pos = m.offset + length(m.match)
        end
        write(io, "$(regex.pattern) $count\n")
    end

    replacements = [
        ('B', "(c|g|t)"),
        ('D', "(a|g|t)"),
        ('H', "(a|c|t)"),
        ('K', "(g|t)"),
        ('M', "(a|c)"),
        ('N', "(a|c|g|t)"),
        ('R', "(a|g)"),
        ('S', "(c|t)"),
        ('V', "(a|c|g)"),
        ('W', "(a|t)"),
        ('Y', "(c|t)"),
    ]

    modified_len = 0
    for ch in b.seq
        replaced = false
        for (from_char, to_str) in replacements
            if ch == from_char
                modified_len += length(to_str)
                replaced = true
                break
            end
        end
        if !replaced
            modified_len += 1
        end
    end

    write(io, "\n")
    write(io, "$(b.ilen)\n")
    write(io, "$(b.clen)\n")
    write(io, "$modified_len\n")
end

function checksum(b::RegexDna)::UInt32
    return Helper.checksum(String(take!(b.io)))
end

File: nim/src/benchmarks/regexdna.nim
import std/[strutils, strformat, sequtils, algorithm, math, tables, re]
import ../benchmark
import ../helper
import fasta

type
  RegexDna* = ref object of Benchmark
    seqData: string
    ilen: int
    clen: int
    resultStr: string
    compiledPatterns: seq[Regex]

proc newRegexDna(): Benchmark =
  RegexDna()

method name(self: RegexDna): string = "CLBG::RegexDna"

const patternStrings = [
  "agggtaaa|tttaccct",
  "[cgt]gggtaaa|tttaccc[acg]",
  "a[act]ggtaaa|tttacc[agt]t",
  "ag[act]gtaaa|tttac[agt]ct",
  "agg[act]taaa|ttta[agt]cct",
  "aggg[acg]aaa|ttt[cgt]ccct",
  "agggt[cgt]aa|tt[acg]accct",
  "agggta[cgt]a|t[acg]taccct",
  "agggtaa[cgt]|[acg]ttaccct"
]

method prepare(self: RegexDna) =

  let fastaInst = cast[Fasta](newFasta())
  fastaInst.n = self.config_val("n")
  fastaInst.prepare()
  fastaInst.run(0)
  let res = fastaInst.get_result()

  self.seqData = ""
  self.ilen = 0

  for line in res.splitLines():
    self.ilen += line.len + 1
    if line.len > 0 and line[0] != '>':
      self.seqData.add(line)

  self.ilen -= 1
  self.clen = self.seqData.len

  self.compiledPatterns = newSeq[Regex]()
  for patternStr in patternStrings:
    try:
      let regex = re(patternStr)
      self.compiledPatterns.add(regex)
    except:
      echo "Error compiling regex: ", patternStr

  self.resultStr = ""

method run(self: RegexDna, iteration_id: int) =
  for i, pattern in self.compiledPatterns.pairs:
    var count = 0
    var pos = 0

    while pos < self.seqData.len:
      let (first, last) = self.seqData.findBounds(pattern, pos)
      if first < 0:
        break
      inc count
      pos = last + 1

    self.resultStr.add(patternStrings[i] & " " & $count & "\n")

  const replacements = [
    ('B', "(c|g|t)"),
    ('D', "(a|g|t)"),
    ('H', "(a|c|t)"),
    ('K', "(g|t)"),
    ('M', "(a|c)"),
    ('N', "(a|c|g|t)"),
    ('R', "(a|g)"),
    ('S', "(c|t)"),
    ('V', "(a|c|g)"),
    ('W', "(a|t)"),
    ('Y', "(c|t)")
  ]

  var seq2 = newStringOfCap(self.seqData.len * 9)

  for c in self.seqData:
    var replaced = false
    for (fromChar, toStr) in replacements:
      if c == fromChar:
        seq2.add(toStr)
        replaced = true
        break

    if not replaced:
      seq2.add(c)

  self.resultStr.add("\n")
  self.resultStr.add($self.ilen & "\n")
  self.resultStr.add($self.clen & "\n")
  self.resultStr.add($seq2.len & "\n")

method checksum(self: RegexDna): uint32 =
  checksum(self.resultStr)

registerBenchmark("CLBG::RegexDna", newRegexDna)
